src/strings/ops.c

```diff
@@ -877,6 +882,42 @@ static MVMint64 string_equal_at_ignore_case(MVMThreadContext *tc, MVMString *Hay
         return H_graphs + H_expansion - H_offset >= n_fc_graphs  ? 1 : 0;
     return 0;
 }
+static void knuth_morris_pratt_process_pattern (MVMThreadContext *tc, MVMString *pat, MVMGrapheme32 next[], MVMStringIndex pat_graphs) {
+    MVMint64 i = 0;
+    MVMint64 j = next[0] = -1;
+    while (i < pat_graphs) {
+        if (j == -1 || MVM_string_get_grapheme_at_nocheck(tc, pat, i) == MVM_string_get_grapheme_at_nocheck(tc, pat, j)) {
+            i++; j++;
+            next[i] = (i < pat_graphs && MVM_string_get_grapheme_at_nocheck(tc, pat, j) == MVM_string_get_grapheme_at_nocheck(tc, pat, i))
+                ? next[j]
+                : j;
+        }
+        else j = next[j];
+    }
+}
+static MVMint64 knuth_morris_pratt_index (MVMThreadContext *tc, MVMString *pat, MVMString *MVM_text, MVMint64 H_offset) {
+    MVMint64 pat_offset = 0, text_offset = H_offset;
+    MVMStringIndex MVM_text_graphs = MVM_string_graphs(tc, MVM_text);
+    MVMStringIndex pat_graphs      = MVM_string_graphs(tc, pat);
+    MVMGrapheme32 *next = alloca(pat_graphs * sizeof(MVMGrapheme32));
+
+    if (knuth_morris_pratt_max_pattern_length < pat_graphs) {
+        MVM_exception_throw_adhoc(tc, "Tried to search a pattern %i long but can only do %i\n", pat_graphs, knuth_morris_pratt_max_pattern_length);
+    }
```

Comment:
This seems like it'd be an internal error, and not something the user should be able to cause (and therefore catch). So an exception seems like it's not the right thing, and maybe a form of assert() instead.